Provide Activity scope for @EBean

smaugho opened the pull request:
This PR adds the scope Activity and scope Fragment to the `@EBean` annotation, solving #1488  

Please note, that this PR is for a revision, cause'  it depends on #2202 . Once I get #2202, I should make a rebase, and then this one would be ready to merge. 

But please, have a look at it, and provide feedback if possible. 

Thanks!

WonderCsabo commented on AndroidAnnotations/androidannotations-core/androidannotations-api/src/main/java/org/androidannotations/annotations/EBean.java:
```diff
@@ -110,6 +110,22 @@
 		Default, //
 
+		/**
+		 * A new instance of the bean is created the first time it is needed in an
```
A new instance of the bean is created the first time it is needed in an
Activity scope, it is then retained and the same instance is always returned
from within the same Activity. Different Activities hold different copies of
the bean.

WonderCsabo commented on AndroidAnnotations/androidannotations-core/androidannotations-api/src/main/java/org/androidannotations/annotations/EBean.java:
```diff
@@ -110,6 +110,22 @@
+		Activity,
+
+		/**
+		 * A new instance of the bean is created the first time it is needed in aa
```
A new instance of the bean is created the first time it is needed in a
Fragment scope, it is then retained and the same instance is always returned
from within the same Fragment. Different Fragments hold different copies of
the bean.

WonderCsabo commented on AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/holder/EBeanHolder.java:
```diff
@@ -111,23 +116,30 @@ public void invokeInitInConstructors() {
-		if (hasSingletonScope) {
+		switch (scope) {
+		case Default:
+		case Fragment:
```
How come `Default` and `Fragment` is in the same branch? Maybe i do not understand something?

smaugho replied:
Because they share common code, but note that below it is called `createOverloadedFactoryMethod`, and it is there where a difference between both scopes is set. The fragment scope code is:

```java
    public static TestModel_ getInstance_(Context context, Object rootFragment) {
        if (rootFragment instanceof BeanHolder) {
            BeanHolder beanHolder = ((BeanHolder) rootFragment);
            TestModel_ bean = beanHolder.getBean(TestModel_.class);
            if (bean == null) {
                bean = new TestModel_(context, rootFragment);
                beanHolder.putBean(TestModel_.class, bean);
            }
            return bean;
        }
        return new TestModel_(context, rootFragment);
    }
```

The Default scope only has:

```java
    public static TestModel_ getInstance_(Context context, Object rootFragment) {
        return new TestModel_(context, rootFragment);
    }
```

smaugho commented:
I added all the tests, did also some organization in existing tests, so please have a look to it. Note that the "Cyclic Tests" are using a Fragment now done only for that end. Also added "Scope Tests" and moved some existing tests related to the "Singleton" scope under that one.

smaugho commented:
Ah Important, this PR depends on #2202, it is basically built over that one (so if you want to see the changes introduced only by this PR, compare it not with "develop", but with #2202 branch. I guess that for merging it, we should firstly do #2202, and afterward this one.

WonderCsabo commented:
@smaugho this should be rebased, right?

smaugho commented:
It should be rebased. But I would wait to merging #2202 , since it depends on it as well, so that I have to rebase only once.

dodgex commented:
#2202 has been merged. could you please rebase and ensure the conflicts are gone? thanks :)

dodgex commented:
@WonderCsabo you still have requested changes here. are they resolved but you missed to update your review or are the changes still pending?

WonderCsabo commented:
@dodgex i think they were resolved, waiting for rebase now. :)

dodgex commented:
@WonderCsabo okay, after rebase approve the PR and then I'll review.

smaugho commented:
Gonna rebase now 👍
